### DISCUSSION TITLE
How to recover from deadlock at cursor.fetchall() ?

### DISCUSSION THREAD

**Discussion by nsmcan** [mkleehammer/pyodbc#1379] [Q&A]
On a SQL Server, sometimes a SELECT query completes successfully by cursor.execute() but gets deadlocked and killed at cursor.fetchall():

```
self.cursor.execute(sql, *values)
self.rowcount = self.cursor.rowcount
results = self.cursor.fetchall()
```

Exception:
```
    results = self.cursor.fetchall()
pyodbc.Error: ('40001', '[40001] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Transaction (Process ID 118) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction. (1205) (SQLFetch)')
```

What would be a proper way to recover there - re-run cursor.fetchall() or re-run cursor.execute() followed by cursor.fetchall()?

**Comment by mkleehammer** [comment]
The entire transaction was aborted, so you'll need to rerun execute.
